### PR TITLE
spiffs fixes

### PR DESCRIPTION
**Pull request by yamt**
## Summary
Make spiffs work on images built with the following tools:
    * https://github.com/igrr/mkspiffs
    * spiffsgen.py from ESP-IDF
      
## Impact
this breaks on-flash compatibility with the previous versions
(CONFIG_SPIFFS_COMPAT_OLD_NUTTX=y is supposed to restore the compatibility)

## Testing
tested locally with esp32 devkitc and the above mentioned tools

### CODE REVIEW

**Comment by btashton** [comment]
Have you run this against the test tools in the sim? Last time there were some changes that broke the tests.

Also we need to figure out what to do about backwards compatibility. I don't think that breaking people who are upgrading is going to be acceptable.

Can you describe what exactly changed in the format? Maybe we can have a compatibility config flag.

**Comment by yamt** [comment]
> Have you run this against the test tools in the sim? Last time there were some changes that broke the tests.

which test tools?

> Also we need to figure out what to do about backwards compatibility. I don't think that breaking people who are upgrading is going to be acceptable.

is anyone using spiffs?
if yes, i want to know how they are using it.
the current code doesn't seem to work with mkspiffs or spiffsgen. (that's why i made this PR)
but i suspect there is a tool i'm not aware of and it's compatible with the current nuttx implementation.

> Can you describe what exactly changed in the format? Maybe we can have a compatibility config flag.

* the leading '/' in the object names
* the alignment in spiffs_pgobj_ndxheader_s

**Comment by btashton** [comment]
This is the test config in the sim

https://www.github.com/apache/incubator-nuttx/tree/master/boards%2Fsim%2Fsim%2Fsim%2Fconfigs%2Fspiffs%2Fdefconfig

**Comment by yamt** [comment]
> This is the test config in the sim
> 
> https://www.github.com/apache/incubator-nuttx/tree/master/boards%2Fsim%2Fsim%2Fsim%2Fconfigs%2Fspiffs%2Fdefconfig

thank you. i run it with and without this PR. the output was identical.

File: fs/spiffs/src/spiffs_vfs.c [hunk 1]
```diff
@@ -408,7 +416,8 @@ static int spiffs_open(FAR struct file *filep, FAR const char *relpath,
 
   /* Check of the file object already exists */
 
-  ret = spiffs_find_objhdr_pgndx(fs, (FAR const uint8_t *)relpath, &pgndx);
+  ret = spiffs_find_objhdr_pgndx(fs, (FAR const uint8_t *)spiffs_relpath,
```
Comment (xiaoxiang781216) [inline fs/spiffs/src/spiffs_vfs.c:419]: can we enhance spiffs_find_objhdr_pgndx and spiffs_fobj_create to handle with or without the leading '/' correclty?

Comment (yamt) [reply fs/spiffs/src/spiffs_vfs.c:419]: if you mean depends on a config, we can. but i feel spiffs_vfs.c is a better place to handle nuttx-specific glitches like this.

if you mean a runtime detection, i feel it's overkill.

File: fs/spiffs/src/spiffs_vfs.c [hunk 2]
```diff
@@ -330,7 +330,14 @@ static int spiffs_readdir_callback(FAR struct spiffs_s *fs,
       DEBUGASSERT(dir != NULL);
       entryp = &dir->fd_dir;
 
-      strncpy(entryp->d_name, (FAR char *)objhdr.name, NAME_MAX + 1);
+      /* Skip the leading '/'. */
+
+      if (objhdr.name[0] != '/')
```
Comment (xiaoxiang781216) [inline fs/spiffs/src/spiffs_vfs.c:335]: how about:
if (objhdr.name[0] != '/')
  {
    strncpy(entryp->d_name, (FAR char *)objhdr.name , NAME_MAX + 1);
  }
else
  {
    strncpy(entryp->d_name, (FAR char *)objhdr.name + 1, NAME_MAX + 1);
  }

File: fs/spiffs/src/spiffs_vfs.c (hunk 1 above)
Comment (xiaoxiang781216) [reply fs/spiffs/src/spiffs_vfs.c:419]: The change add local variable and memcpy which increase the stack usage and reduce the performance. And every caller of spiffs_find_objhdr_pgndx/spiffs_fobj_create need to do the same thing repeatly. If we handle the leading '/' inside spiffs_find_objhdr_pgndx/spiffs_fobj_create, we can avoid the local variable and only change one place.

File: fs/spiffs/src/spiffs_vfs.c (hunk 2 above)
Comment (yamt) [reply fs/spiffs/src/spiffs_vfs.c:335]: do you have specific reasons to make this that smart?
generally i don't think it's a good idea to put this kind of smart things unless absolutely necessary.

File: fs/spiffs/src/spiffs_vfs.c (hunk 1 above)
Comment (yamt) [reply fs/spiffs/src/spiffs_vfs.c:419]: one place? you mean three?

i can change it if you feel strongly.
but i guess it isn't the biggest issue for this PR.
the incompatibility with older nuttx is.

File: fs/spiffs/src/spiffs_vfs.c (hunk 2 above)
Comment (xiaoxiang781216) [reply fs/spiffs/src/spiffs_vfs.c:335]: Sure, it's fine if you confirm the lead '/' always exist.

File: fs/spiffs/src/spiffs_vfs.c (hunk 1 above)
Comment (xiaoxiang781216) [reply fs/spiffs/src/spiffs_vfs.c:419]: > one place? you mean three?
> 

we need change at least spiffs_open, spiffs_unlink, spiffs_rename and spiffs_stat with the current patch. On the other hand, if we enhance spiffs_find_objhdr_pgndx to support the name without leading '\', all spiffs_xxx don't need change.
And these patches also don't need anymore:
fs/spiffs/Kconfig: Update the description of SPIFFS_NAME_MAX
spiffs: Fix statfs f_namelen

> i can change it if you feel strongly.
> but i guess it isn't the biggest issue for this PR.
> the incompatibility with older nuttx is.

Yes, the padding added inside spiffs_pgobj_ndxheader_s is more critical than others. We need reference the offical spiffs to see how they define this structure.

Comment (yamt) [reply fs/spiffs/src/spiffs_vfs.c:419]: the structure with this PR should be compatible with:
* https://github.com/pellepl/spiffs (the version used by mkspiffs. i dunno if 0.4.0 changed it or not)
* https://github.com/igrr/mkspiffs
* ESP-IDF spiffsgen.py

Comment (yamt) [reply fs/spiffs/src/spiffs_vfs.c:419]: > And these patches also don't need anymore:
> fs/spiffs/Kconfig: Update the description of SPIFFS_NAME_MAX
> spiffs: Fix statfs f_namelen

i'm not sure about this point.
do you prefer to keep SPIFFS_NAME_MAX not including the leading '/'?
in that case, a user need to specify SPIFFS_NAME_MAX+1 for tools like mkspiffs.

Comment (xiaoxiang781216) [reply fs/spiffs/src/spiffs_vfs.c:419]: > the structure with this PR should be compatible with:
> 
> * https://github.com/pellepl/spiffs (the version used by mkspiffs. i dunno if 0.4.0 changed it or not)
> * https://github.com/igrr/mkspiffs
> * ESP-IDF spiffsgen.py

If so, we need change the NuttX version to confirm the standard as soon as possible.

> > And these patches also don't need anymore:
> > fs/spiffs/Kconfig: Update the description of SPIFFS_NAME_MAX
> > spiffs: Fix statfs f_namelen
> 
> i'm not sure about this point.
> do you prefer to keep SPIFFS_NAME_MAX not including the leading '/'?
> in that case, a user need to specify SPIFFS_NAME_MAX+1 for tools like mkspiffs.

From your patch, '/' is always inserted before the relpath. If spiffs_find_objhdr_pgndx can handle relpath without the leading '/' correctly. Then we don't need reserve one space for '/'?

Comment (yamt) [reply fs/spiffs/src/spiffs_vfs.c]: done

**Comment by yamt** [comment]
i introduced a Kconfig for the compatibility with older nuttx versions.
i haven't tested it as i don't know how to create images for them.

**Comment by yamt** [comment]
> > This is the test config in the sim
> > https://www.github.com/apache/incubator-nuttx/tree/master/boards%2Fsim%2Fsim%2Fsim%2Fconfigs%2Fspiffs%2Fdefconfig
> 
> thank you. i run it with and without this PR. the output was identical.

btw, i couldn't find where the spiffs filesystem is initialized for the config.
is the test assuming that the initial state of RAM MTD is a valid spiffs?
is it the right thing to do?

File: fs/spiffs/src/spiffs_vfs.c (hunk 1 above)
Comment (yamt) [reply fs/spiffs/src/spiffs_vfs.c]: > > the structure with this PR should be compatible with:
> > 
> > * https://github.com/pellepl/spiffs (the version used by mkspiffs. i dunno if 0.4.0 changed it or not)
> > * https://github.com/igrr/mkspiffs
> > * ESP-IDF spiffsgen.py
> 
> If so, we need change the NuttX version to confirm the standard as soon as possible.

it's what this PR does.

> 
> > > And these patches also don't need anymore:
> > > fs/spiffs/Kconfig: Update the description of SPIFFS_NAME_MAX
> > > spiffs: Fix statfs f_namelen
> > 
> > 
> > i'm not sure about this point.
> > do you prefer to keep SPIFFS_NAME_MAX not including the leading '/'?
> > in that case, a user need to specify SPIFFS_NAME_MAX+1 for tools like mkspiffs.
> 
> From your patch, '/' is always inserted before the relpath. If spiffs_find_objhdr_pgndx can handle relpath without the leading '/' correctly. Then we don't need reserve one space for '/'?

regardless of what nuttx do, those tools take a length including the leading slash.

in the latest push, i made it conditional for people who prefer compatibility with older nuttx.

**Comment by xiaoxiang781216** [comment]
@btashton and @patacongo should we keep the compatibility(mistake) layer?

**Comment by yamt** [comment]
after reading https://github.com/pellepl/spiffs code a bit, i feel the leading slash might not be essential in the on-flash structure.
it's probably just how mkspiffs and spiffsgen.py generate an image from a directory.
i guess nothing prevents a spiffs-using software from using SPIFFS_open (the api in the original spiffs)
with a name without slashes.

on the other hand, the alignement things seems essential.

i guess it makes more sense to have two separate Kconfig for them.

**Comment by yamt** [comment]
> i guess it makes more sense to have two separate Kconfig for them.

i separated them

**Comment by yamt** [comment]
does any of you still have any concerns?

**Review by xiaoxiang781216 — APPROVED** [review]
Look good to me now.

**Review by jerpelea — APPROVED** [review]

**Comment by yamt** [comment]
> Look good to me now. But @yamt could you reorg your pathset to more reasonable?
> 
>     1. Merge "fs/spiffs/Kconfig: Add SPIFFS_COMPAT_OLD_NUTTX" and "spiffs_pgobj_ndxheader_s: Add a missing alignment"

done

>     2. Merge "spiffs: Prefix filenames with '/' as other implmenetations do" and "spiffs: Suggest SPIFFS_LEADING_SLASH=n for better compatibility"

done

>     3. Merge "spiffs: Rename the upsteam readme to README-spiffs.md" and "spiffs: Document how to generate images"

i don't want to do this because it makes a diff difficult to read.

**Comment by xiaoxiang781216** [comment]
Ok.

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**